federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,353 @@
+				// When new cluster becomes available process all the namespaces again.
+				nc.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(ClusterAvailableDelay))
+			},
+		},
+	)
+
+	// Federated updeater along with Create/Update/Delete operations.
+	nc.federatedUpdater = util.NewFederatedUpdater(nc.namespaceFederatedInformer,
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace, _ := obj.(*api_v1.Namespace)
+			_, err := client.Core().Namespaces().Create(namespace)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace, _ := obj.(*api_v1.Namespace)
+			_, err := client.Core().Namespaces().Update(namespace)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace, _ := obj.(*api_v1.Namespace)
+			err := client.Core().Namespaces().Delete(namespace.Name, &api.DeleteOptions{})
+			return err
+		})
+	return nc
+}
+
+func (nc *NamespaceController) Start() {
+	nc.namespaceInformerController.Run(nc.namespaceStopChan)
+	nc.namespaceFederatedInformer.Start()
+
+	// Handle cluster available/unavailable events.
+	go func() {
+		for {
+			select {
+			// All messages on this chanel mean that all namespaces should be reconciliated.
+			case <-nc.clusterDeliverer.GetTargetChanel():
+				nc.reconcileNamespacesOnClusterChange()
+			case <-nc.clusterReconcileStopChan:
+				return
+			}
+		}
+	}()
+
+	// Handle namespace reconciliation.
+	go func() {
+		for {
+			select {
+			case item := <-nc.namespaceDeliverer.GetTargetChanel():
+				ni := item.Value.(*namespaceItem)
+				nc.reconcileNamespace(ni.namespace, ni.trial)
+			case <-nc.namespaceReconcileStopChan:
+				return
+			}
+		}
+	}()
+}
+
+func (nc *NamespaceController) Stop() {
+	nc.namespaceFederatedInformer.Stop()
+	close(nc.namespaceStopChan)
+	close(nc.clusterReconcileStopChan)
+	close(nc.namespaceReconcileStopChan)
+}
+
+func (nc *NamespaceController) deliverNamespaceObj(obj interface{}, delay time.Duration, trial int64) {
+	namespace, ok := obj.(*api.Namespace)
+	if ok {
+		nc.deliverNamespace(namespace.Name, delay, trial)
+	}
+}
+
+func (nc *NamespaceController) deliverNamespace(namespace string, delay time.Duration, trial int64) {
+	nc.namespaceDeliverer.DeliverAt(
+		namespace,
+		&namespaceItem{
+			namespace: namespace,
+			trial:     trial},
+		time.Now().Add(delay))
+}
+
+func (nc *NamespaceController) isSynced() bool {
+	clusters, err := nc.namespaceFederatedInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get ready clusters: %v", err)
+		return false
+	}
+	if !nc.namespaceFederatedInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+	if !nc.namespaceFederatedInformer.GetTargetStore().ClustersSynced(clusters) {
+		return false
+	}
+	return true
+}
+
+func (nc *NamespaceController) reconcileNamespacesOnClusterChange() {
+	if !nc.isSynced() {
```

Comment:
Can you please add a comment what exactly this checks? IIUC, if there is a "new cluster" that is not yet fully synced, then we will delay the reconciliation. Is it the only case when we are delaying, or are there any others? It would be good to have a comment on it here.